### PR TITLE
2nd attempt: py3 and parallelism

### PR DESCRIPTION
**Pull request by sbailey**
This is a second attempt at PR #5 to bring in python 3 compatibility and new features for parallelism implemented by DESI.  @timahutchinson caught that some of his changes for template groups had been inadvertently reverted, so this branch reverts those reverts.

Changes include:
- updates for python 3.5, while still retaining python 2.7 compatibility. The vast majority of these changes are to add parentheses to print and change xrange to range.
- New optional multiprocessing-based parallelism in zfinder.py.
- a new etc/redmonster.module template for desiInstall; should have no impact on SDSS

Please review carefully, since our multi-repo commit history is rather entangled and this touches a lot of files.

### CODE REVIEW

File: python/redmonster/physics/zfinder.py [hunk 1]
```diff
@@ -63,198 +130,176 @@ def read_template(self):
...
             self.zwarning[i] = int(self.zwarning[i]) | \
                     int(temp_zwarning[i,minpos[0],minpos[1]])
         zchi2arr = n.reshape(zchi2arr, (specs.shape[0],) + self.origshape[:-1] +
                              (num_z,) )
         bestl = n.where(zchi2arr == n.min(zchi2arr))[-1][0]
-        if bounds_set:
-            thisz = ((10**(specloglam[0]))/self.tempwave[bestl+zminpix])-1
-        else:
-            thisz = ((10**(specloglam[0]))/self.tempwave[bestl])-1
+        thisz = ((10**(specloglam[0]))/self.tempwave[bestl+zminpix])-1
```
Comment (sbailey) [inline python/redmonster/physics/zfinder.py:285]: I'm not sure why this `bound_set` logic was changed on the desi side.  @timahutchinson please review if this is needed and we can fix.

File: python/redmonster/physics/zfinder.py [hunk 2]
```diff
@@ -4,42 +4,109 @@
...
+    f = (a[ii]!=0)*b[ii]/(a[ii]+(a[ii]==0))
+    zchi2arr = chi2_0 - a[ii]*f**2 # is this true ?????
+
+    zwarning=n.zeros((num_z))
+    zchi2arr[f<0] = chi2_0
+    zwarning[f<0] = flag_val_neg_model
+    return j,zchi2arr,zwarning
+
+
+
```
Comment (sbailey) [inline python/redmonster/physics/zfinder.py:86]: Note that the changes in Zfinder to add multiprocessing parallelism are the most substantial intended organizational change.  The results should be the same as before, just faster.  i.e. this is a place to test with the sdss data.

File: python/redmonster/physics/zfinder.py [hunk 3]
```diff
@@ -63,198 +130,176 @@ def read_template(self):
...
             if (plate is not None) & (mjd is not None) & (fiberid is not None):
                 write_chi2arr(plate, mjd, fiberid, self.zchi2arr)
             else:
-                print 'Plate/mjd/fiberid not given - unable to write chi2 file!'
+                print('WARNING Plate/mjd/fiberid not given - unable to write chi2 file!')
         else:
-            print 'Not writing chi2'
-
+            #print 'INFO Not writing chi2'
+            pass
```
Comment (sbailey) [inline python/redmonster/physics/zfinder.py]: This might be a leftover edit; do you want this print?

File: python/redmonster/physics/zfinder.py [hunk 4]
```diff
@@ -4,42 +4,109 @@
...
+
+
+
 class ZFinder:
-    
-    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None):
+    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None, nproc=1):
         self.fname = fname
-        if type(group) == 'list':
+        if type(group) == list:
```
Comment (sbailey) [inline python/redmonster/physics/zfinder.py:90]: Note: bug fix for `'list'` vs. `list` (no quotes), and also supporting group as an integer and not just a list (convenient, but I'm not sure if that fits within your big picture for template groups usage).

File: python/redmonster/physics/zfinder.py [hunk 5]
```diff
@@ -4,42 +4,109 @@
...
             self.group = group
+        elif type(group) == int:
+            self.group = [group,]
         else:
             self.group = eval(group)
-        self.npoly = npoly if npoly else 4
+        self.npoly = npoly if npoly is not None else 4
```
Comment (sbailey) [inline python/redmonster/physics/zfinder.py:96]: Note: fixes case for `npoly=0`, which would have been set to `self.npoly = 4` with original code.  This might explain some odd performance vs. npoly seen my Tim and Kyle if they thought they were testing npoly=0 but were actually getting npoly=4.

**Review by sbailey — COMMENTED** [review]
I added some inline comments for areas to pay special attention to.

**Comment by sbailey** [comment]
I added some inline comments for areas to look at carefully.  I'll also note that I tested this with desi sims using python 2.7 and 3.5, but there is a lot of sdss-specific and special-purpose code that has not been tested.  The changes there are based upon 2to3 recommendations.

File: python/redmonster/physics/zfinder.py (hunk 3 above)
Comment (timahutchinson) [reply python/redmonster/physics/zfinder.py]: I believe this should still print.

File: python/redmonster/physics/zfinder.py (hunk 1 above)
Comment (timahutchinson) [reply python/redmonster/physics/zfinder.py:285]: It looks like this will still work, as the same logic is still there, but was moved elsewhere.  I don't understand why, though.

File: python/redmonster/physics/zfinder.py (hunk 5 above)
Comment (timahutchinson) [reply python/redmonster/physics/zfinder.py:96]: You're right - this was a bug.

File: python/redmonster/physics/zfinder.py (hunk 4 above)
Comment (timahutchinson) [reply python/redmonster/physics/zfinder.py:90]: In general, template groups should always be passed in as lists, but it's probably good to have it support int as well.

File: python/redmonster/_version.py [hunk 6]
```diff
@@ -1 +1 @@
-__version__ = 'v1.1.0'
+__version__ = '1.1.1.dev3'
```
Comment (timahutchinson) [inline python/redmonster/_version.py]: This should remain as is in my version for consistency.

File: python/redmonster/physics/zfinder.py [hunk 7]
```diff
@@ -4,42 +4,109 @@
...
         self.zmin = float(zmin)
         self.zmax = float(zmax)
+        self.nproc = nproc
         self.pixoffset = None
         self.zchi2arr = None
+        self.nproc=nproc
```
Comment (timahutchinson) [inline python/redmonster/physics/zfinder.py]: This is being set twice - here and in line 99.

File: python/redmonster/physics/zpicker2.py [hunk 8]
```diff
@@ -107,7 +108,7 @@ def classify_obj(self, zfindobjs, zfitobjs, flags, rchi2threshold=0.005):
             npixsteptuple = ()
             fstuple = ()
             # Catch spectra that are all 0's and return null result
-            if len(n.where(self.flux[ifiber] != 0)[0]) == 0 or len(n.where(self.ivar[ifiber] != 0)[0]) == 0:
+            if len(n.where(self.flux[0] != 0)[0]) == 0:
```
Comment (timahutchinson) [inline python/redmonster/physics/zpicker2.py]: This needs to be reverted to the way it is in my copy. With this edit, it will not properly handle spectra that are all zeros if they aren't the first spectra in the array. Also, why was the check for all(ivar) == 0 removed?

File: python/redmonster/physics/zpicker2.py [hunk 9]
```diff
@@ -138,11 +139,17 @@ def classify_obj(self, zfindobjs, zfitobjs, flags, rchi2threshold=0.005):
...
-                                    (self.dof[ifiber] - zfindobjs[itemp].npoly))
+                            #print "%r" % e
+                            #fibermins.append( \
+                            #        n.max(zfitobjs[itemp].chi2vals[ifiber]) / \
+                            #        (self.dof[ifiber] - zfindobjs[itemp].npoly))
+
+                            # this is ok, might be no solution
+                            #print "WARNING no z fitted for fiber #%d, class #%d, zid #%d"%(ifiber,itemp,imin)
+
+                            fibermins.append(100000.)
```
Comment (timahutchinson) [inline python/redmonster/physics/zpicker2.py]: Why was this changed? It should be left as is.

Comment (sbailey) [reply python/redmonster/physics/zpicker2.py]: @julienguy changed that with the comment

> a bug has also been fixed when an IndexError is raised in the situation where no refined redshift is found for a given template class (I checked the best redshift/class from the other template classes is still return correctly).

If I interpret that correctly, the problem arises when `zfitobjs[itemp].chi2vals` is an empty list, which will cause `n.max(zfitobjs[itemp].chi2vals[ifiber])` to raise a `ValueError`.  The proposed fix of `fibermins.append(100000.)` also seems pretty hacky, but it is robust... :)  Thoughts on this case of no refined redshifts found for a given template class?

File: python/redmonster/physics/zfinder.py (hunk 4 above)
Comment (sbailey) [reply python/redmonster/physics/zfinder.py:90]: Is the proper use of template groups documented?  Earlier desispec code was calling redmonster with an integer group which then broke when this changed to a list.  I added the support for an integer for convenience to not have to update the desispec code immediately.  We can migrate the desispec code too, but it would be helpful to understand how these are supposed to be used so that the desispec code can be updated correctly.

File: python/redmonster/physics/zfinder.py (hunk 3 above)
Comment (sbailey) [reply python/redmonster/physics/zfinder.py]: updated.

File: python/redmonster/_version.py (hunk 6 above)
Comment (sbailey) [reply python/redmonster/_version.py]: I put back the 'v', but the version number should be different from 'v1.1.0' since the code no longer matches that tag.

File: python/redmonster/physics/zpicker2.py (hunk 8 above)
Comment (sbailey) [reply python/redmonster/physics/zpicker2.py]: I'm not sure why that was changed.  I put it back, though using a different method to check for all zeros: `if n.all(self.flux[ifiber] == 0.0) or n.all(self.ivar[ifiber] == 0.0):`

File: python/redmonster/_version.py (hunk 6 above)
Comment (timahutchinson) [reply python/redmonster/_version.py]: I agree.  This will be fine for this PR, and I can address consistency when it's tagged.

File: python/redmonster/physics/zpicker2.py (hunk 8 above)
Comment (timahutchinson) [reply python/redmonster/physics/zpicker2.py]: Great!

File: python/redmonster/physics/zpicker2.py (hunk 9 above)
Comment (timahutchinson) [reply python/redmonster/physics/zpicker2.py]: I remember this issue arising.  I seem to recall implementing a fix downstream (to deal with an empty chi2vals list), but I can't remember for sure now.  Let me look into it tomorrow and come back to this.

File: python/redmonster/physics/zfinder.py (hunk 4 above)
Comment (timahutchinson) [reply python/redmonster/physics/zfinder.py:90]: It's not - I need to do that.  Template groups were originally implemented as integers, as each template could be assigned to a single group.  I didn't feel like this was a complete solution to the problem, though, since the objects that are being classified aren't discrete, but exist on a continuum of sorts.  For example, maybe you want galaxy templates and quasar templates to flag against each other, but you don't want emission line galaxies to flag against galaxy or quasar.  There wasn't a way to have this happen with single groups, but assigned galaxy to group 1, ELG to groups [1,2], and quasar to group 2.  In order to do so, I sort of standardized that all groups are lists, and code that assumes that.

File: python/redmonster/physics/zpicker2.py (hunk 9 above)
Comment (timahutchinson) [reply python/redmonster/physics/zpicker2.py]: Okay, so looking into this a bit further - it looks like it should be fixed by lines 171 and 188 in zfitter.py , unless there's some other case where an empty list is sneaking through.  I don't know if Julien was seeing this crash before or after I added this.

Regardless, maybe a better way to handle it would be to simply have a try/except block here - it'll try to add the maximum value (as in my version), and if a ValueError is raised, then fall back on the clunkier solution of just adding a large chi2 value.

File: python/redmonster/physics/zfinder.py (hunk 7 above)
Comment (sbailey) [reply python/redmonster/physics/zfinder.py]: fixed

File: python/redmonster/physics/zpicker2.py (hunk 9 above)
Comment (sbailey) [reply python/redmonster/physics/zpicker2.py]: I updated this to check for 0-length list before taking the max, falling back to only using a large value if needed.

**Comment by sbailey** [comment]
I think I've addressed the comments now.  Anything else?

**Comment by timahutchinson** [comment]
I think that fixes all the issues I saw - I'll merge the pull request and test it tomorrow morning.
